### ISSUE TITLE
More NotWordBoundaryAscii woes

### ISSUE THREAD

**Issue by SeanRBurton** [rust-lang-nursery/regex#264]
I'm sorry to say that another version of issue #241 has reared its ugly head.

Running `Group { e: NotWordBoundaryAscii, i: Some(1), name: None }` against `"\u{28f3e}"` in the same way as in that issue gives the match `None` with the DFA, but `Some((0, 0))` with the NFA.

**Comment by BurntSushi** [comment]
A smallerish repro:

```rust
extern crate regex;

fn main() {
    let re0 = regex::Regex::new(r"(?-u)\B").unwrap();
    let re1 = regex::Regex::new(r"(?-u)(\B)").unwrap();
    let s = "\u{28f3e}";
    assert!(re0.is_match(s));
    assert!(re1.is_match(s));
    assert!(re0.find(s).is_some());
    assert!(re1.find(s).is_some());
    assert!(re0.captures(s).is_some());
    assert!(re1.captures(s).is_some());
}
```

Even more interesting is that both

**Comment by BurntSushi** [comment]
The above actually fails on the last assert. My guess is that something about the DFA -> NFA interaction is bunk, because the presence of the capture group *and* the call to `captures` is what will force the exec engine to run the NFA after the DFA.

**Comment by BurntSushi** [comment]
Indeed. The code was trying to be cute about setting the end of the haystack. In this case, it was adding `1` unconditionally to the end of the match location when in fact it needs to add a number of bytes equal to the next encoded codepoint.
